builtin/logical/aws/rotation_test.go

```diff
@@ -0,0 +1,196 @@
+		creds []credToInsert
+	}{
+		{
+			name: "refresh one",
+			creds: []credToInsert{
+				{
+					config: staticRoleConfig{
+						Name:           "test",
+						Username:       "jane-doe",
+						RotationPeriod: 2 * time.Second,
+					},
+					changed: true,
+				},
+			},
+		},
+		{
+			name: "refresh none",
+			creds: []credToInsert{
+				{
+					config: staticRoleConfig{
+						Name:           "test",
+						Username:       "jane-doe",
+						RotationPeriod: 1 * time.Minute,
+					},
+					changed: false,
+				},
+			},
+		},
+		{
+			name: "refresh one of two",
+			creds: []credToInsert{
+				{
+					config: staticRoleConfig{
+						Name:           "test",
+						Username:       "jane-doe",
+						RotationPeriod: 1 * time.Minute,
+					},
+					changed: false,
+				},
+				{
+					config: staticRoleConfig{
+						Name:           "toast",
+						Username:       "john-doe",
+						RotationPeriod: 1 * time.Second,
+					},
+					changed: true,
+				},
+			},
+		},
+	}
+
+	// wrap our cred generating function so we don't have to handle the error _every time_.
+	// we pass in the T so that we have the correct subtest T for each call.
+	randStr := func(len int, t *testing.T) *string {
+		str, err := credsutil.RandomAlphaNumeric(len, false)
+		if err != nil {
+			t.Fatalf("couldn't generate a random string for a key: %s", err)
+		}
+		return aws.String(str)
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			config := logical.TestBackendConfig()
+			config.StorageView = &logical.InmemStorage{}
+
+			b := Backend()
+			err := b.initQueue(context.Background(), nil)
+			if err != nil {
+				t.Fatalf("couldn't initialize queue: %s", err)
+			}
+
+			// this means the creds will be the same for every user, but that's okay
+			// since what we care about is whether they changed on a single-user basis.
+			miam, err := awsutil.NewMockIAM(
+				// blank list for existing user
+				awsutil.WithListAccessKeysOutput(&iam.ListAccessKeysOutput{
+					AccessKeyMetadata: []*iam.AccessKeyMetadata{
+						{},
+					},
+				}),
+				// initial key to store
+				awsutil.WithCreateAccessKeyOutput(&iam.CreateAccessKeyOutput{
+					AccessKey: &iam.AccessKey{
+						AccessKeyId:     randStr(16, t),
+						SecretAccessKey: randStr(16, t),
+					},
+				}),
+			)(nil)
+
+			// insert all our creds
+			for i, cred := range c.creds {
+
+				if err != nil {
+					t.Fatalf("couldn't initialze mock IAM handler: %s", err)
+				}
+				b.iamClient = miam
+
+				err = b.createCredential(context.Background(), config.StorageView, cred.config)
+				if err != nil {
+					t.Fatalf("couldn't insert credential %d: %s", i, err)
+				}
+
+				item := &queue.Item{
+					Key:      cred.config.Name,
+					Value:    cred.config,
+					Priority: time.Now().Add(cred.config.RotationPeriod).Unix(),
+				}
+				err = b.credRotationQueue.Push(item)
+				if err != nil {
+					t.Fatalf("couldn't push item onto queue: %s", err)
+				}
+			}
+
+			time.Sleep(5 * time.Second)
```

Comment:
We need to wait until the creds have expired so refreshing the queue will actually cause rotations - this could theoretically be any length of time we feel confident will be longer than whatever we're setting the duration to in the cases table.

`createCredential` is actually outside the validation checks we run on RoleWrite, so everything could be set to milliseconds, except credentials we want to expire (e.g., expiring credentials are 1ms, we 'sleep' for 2ms, non-expiring credentials are set to 800 years), but times like 1s, 2s, 5s, etc seemed like a good compromise without making the test take forever.

In the 'real world' the queue is checked once a minute.

Reply:
Since the rotation is the feature that we're testing for, I wonder if we could rewrite the test in such a way that we don't have to _wait_ for the creds to expire.
Is it possible to create a queue that contains already expired credentials?

Otherwise, if we had to wait, I'd suggest using something like `assert.Eventually`: https://pkg.go.dev/github.com/stretchr/testify/assert#Eventually

Reply:
That's fair. 

I liked the old version because it was a straightforward test of what is actually going to be happening, but I have added a _time travel factor_ that preexpires credentials, eliminates the `Sleep` and essentially tests the same thing.